README.md
# CombineExtras

Some handy extras to use alongside the Combine/OpenCombine framework.
Including:
- AnySubject

## Usage

### Publisher.log(...)
Use this to log the output and events of a publisher to a lgger with a  custom log message.

    let logger = Logger(subsystem: "com.example.subsystem", category: "main")

    let counting = [1, 2, 3].publisher
        .log(to: logger, prefix: "Counter") { logger, output in
            logger.log("Counter got \(output, privacy: .private) new coin(s)")
        }
        .sink { _ in } receiveValue: { _ in }

    // sample output:
    // 2021-03-16 20:41:10.980102+0200 subsystem[47288:1810254] [main] Counter started
    // 2021-03-16 20:41:10.980252+0200 subsystem[47288:1810254] [main] Counter got 1 new coin(s)
    // 2021-03-16 20:41:10.980361+0200 subsystem[47288:1810254] [main] Counter got 2 new coin(s)
    // 2021-03-16 20:41:10.980451+0200 subsystem[47288:1810254] [main] Counter got 3 new coin(s)
    // 2021-03-16 20:41:10.980578+0200 subsystem[47288:1810254] [main] Counter finished

### URLSession.loggingDataTask(...)
Creates a publisher that wraps a URL session data task for a given URL request,
printing the request and response with one liners to a logger.

    let logger = Logger(subsystem: "com.example.subsystem", category: "main")

    let url = URL(string: "http://example.com")!
    let request = URLRequest(url: url)

    let fetching = URLSession.shared.loggingDataTask(for: request, using: logger, prefix: "Eg.")
    .sink { _ in  } receiveValue: { _ in }
    }

    // sample output:
    // 2021-04-05 10:31:11.056163+0300 subsystem[20738:1225036] [main] Eg. request: GET http://example.com
    // 2021-04-05 10:31:11.061316+0300 subsystem[20738:1225067] [main] Eg. response: 200 http://example.com  27 byte(s)

### AnySubject

    var subject: AnySubject<String, Never>
    
    // variable can contain different subject types:
    subject = PassthroughSubject<String, Never>().eraseToAnySubject()
    subject = CurrentValueSubject<String, Never>("").eraseToAnySubject()
    
    // type hidden
    let resolvesToNil = PassthroughSubject<String, Never>().eraseToAnySubject() as? PassthroughSubject<String, Never>

## Installation

### Swift Package Manager

Add the following dependency to your **Package.swift** file:

```swift
.package(name: "CombineExtras", url: "https://github.com/andybezaire/CombineExtras.git", from: "1.2.0")
```
## License

Authorization is available under the MIT license. See the LICENSE file for more info.


## Credit

Copyright (c) 2021 andybezaire

Created by: Andy Bezaire


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
### Changed
- logging headders only logs the first 20 characters of the value, this avoids long bearer token prints
### Deprecated
### Removed
### Fixed
### Security

## [1.4.0] - 2021-04-05
### Added
- logging data task

## [1.3.3] - 2021-04-03
### Added
- changelog
### Changed
- tidied up package.swift
- added package related files to gitignore

## [1.3.2] - 2021-03-20

## [1.3.1] - 2021-03-20

## [1.3.0] - 2021-03-20

## [1.2.0] - 2021-03-16

## [1.1.0] - 2021-03-13

## [1.0.0] - 2021-03-06


[Unreleased]: https://github.com/andybezaire/CombineExtras/compare/v1.4.0...HEAD
[1.4.0]: https://github.com/andybezaire/CombineExtras/compare/1.3.3...1.4.0
[1.3.3]: https://github.com/andybezaire/CombineExtras/compare/1.3.2...1.3.3
[1.3.2]: https://github.com/andybezaire/CombineExtras/compare/1.3.1...1.3.2
[1.3.1]: https://github.com/andybezaire/CombineExtras/compare/1.3.0...1.3.1
[1.3.0]: https://github.com/andybezaire/CombineExtras/compare/1.2.0...1.3.0
[1.2.0]: https://github.com/andybezaire/CombineExtras/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/andybezaire/CombineExtras/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/andybezaire/CombineExtras/releases/tag/1.0.0


Sources/CombineExtras/Logging/Logging.swift
//
//  Logging.swift
//
//
//  Created by Andy Bezaire on 16.3.2021.
//

import Combine
import os.log

public extension Publisher {
    /// Use this to log the output and events of a publisher to a lgger with a  custom log message.
    ///
    /// Events are logged respectively as:
    /// - `receiveSubscription: <prefix> started`
    /// - `receiveCompletion.failure: <prefix> error: <error>`
    /// - `receiveCompletion.finished: <prefix> finished`
    /// - `receiveCancel: <prefix> cancelled`
    ///
    /// use the logger in the completion to log an appropriate message and optionally select the privacy of the output
    ///
    /// Usage:
    ///
    ///     let logger = Logger(subsystem: "com.example.subsystem", category: "main")
    ///
    ///     let cancel = [1, 2, 3].publisher
    ///         .log(to: logger, prefix: "Counter") { logger, output in
    ///             logger.log("Counter got \(output, privacy: .private) new coin(s)")
    ///         }
    ///         .sink { _ in } receiveValue: { _ in }
    ///
    ///      // sample output:
    ///      // 2021-03-16 20:41:10.980102+0200 subsystem[47288:1810254] [main] Counter started
    ///      // 2021-03-16 20:41:10.980252+0200 subsystem[47288:1810254] [main] Counter got 1 new coin(s)
    ///      // 2021-03-16 20:41:10.980361+0200 subsystem[47288:1810254] [main] Counter got 2 new coin(s)
    ///      // 2021-03-16 20:41:10.980451+0200 subsystem[47288:1810254] [main] Counter got 3 new coin(s)
    ///      // 2021-03-16 20:41:10.980578+0200 subsystem[47288:1810254] [main] Counter finished
    /// - Parameters:
    ///   - logger: the optional logger to log to
    ///   - prefix: the prefixed text before any of the lifecycle methods
    ///   - logCommand: a closure which can be used to log the output appropriately
    /// - Returns: a publisher similar to the upstream publisher
    func log(
        to logger: Logger?,
        prefix: String,
        _ logCommand: ((Logger, Output) -> Void)? = nil
    ) -> AnyPublisher<Output, Failure> {
        guard let logger = logger else { return self.handleEvents().eraseToAnyPublisher() }

        let receiveSubscription: ((Subscription) -> Void) = { _ in
            logger.log("\(prefix, privacy: .public) started")
        }

        let receiveCompletion: ((Subscribers.Completion<Self.Failure>) -> Void) = { completion in
            switch completion {
            case .failure(let error):
                logger.log(
                    level: .error,
                    "\(prefix, privacy: .public) error: \(error.localizedDescription, privacy: .public)"
                )
            case .finished:
                logger.log("\(prefix, privacy: .public) finished")
            }
        }

        let receiveOutput: ((Output) -> Void) = { output in
            _ = logCommand?(logger, output)
        }

        let receiveCancel: (() -> Void) = {
            logger.log("\(prefix, privacy: .public) cancelled")
        }

        return self.handleEvents(
            receiveSubscription: receiveSubscription,
            receiveOutput: receiveOutput,
            receiveCompletion: receiveCompletion,
            receiveCancel: receiveCancel
        )
        .eraseToAnyPublisher()
    }

    /// Use this to log only the output of a publisher to a custom log message.
    ///
    /// Usage:
    ///
    ///     let logger = Logger(subsystem: "com.example.subsystem", category: "main")
    ///
    ///     let cancel = [1, 2, 3].publisher
    ///         .logOutput(to: logger) { logger, output in
    ///             logger.log("Counter got \(output, privacy: .private) new coin(s)")
    ///         }
    ///         .sink { _ in } receiveValue: { _ in }
    ///
    ///      // sample output:
    ///      // 2021-03-16 20:41:10.980252+0200 subsystem[47288:1810254] [main] Counter got 1 new coin(s)
    ///      // 2021-03-16 20:41:10.980361+0200 subsystem[47288:1810254] [main] Counter got 2 new coin(s)
    ///      // 2021-03-16 20:41:10.980451+0200 subsystem[47288:1810254] [main] Counter got 3 new coin(s)
    /// - Parameters:
    ///   - logger: the optional logger to log to
    ///   - logCommand: a closure which can be used to log the output appropriately
    /// - Returns: a publisher similar to the upstream publisher
    func logOutput(
        to logger: Logger?,
        _ logCommand: ((Logger, Output) -> Void)? = nil
    ) -> AnyPublisher<Output, Failure> {
        let receiveOutput = logger
            .map { logger in { output in _ = logCommand?(logger, output) } }
        return self
            .handleEvents(receiveOutput: receiveOutput)
            .eraseToAnyPublisher()
    }

    /// Use this to log the events of a publisher to a logger.
    ///
    /// Events are logged appropriately as:
    /// - `receiveSubscription: <prefix> started`
    /// - `receiveOutput: <prefix> output`
    /// - `receiveCompletion.failure: <prefix> error: <error>`
    /// - `receiveCompletion.finished: <prefix> finished`
    /// - `receiveCancel: <prefix> cancelled`
    ///
    /// use the logger in the completion to log an appropriate message and optionally select the privacy of the output
    ///
    /// Usage:
    ///
    ///     let logger = Logger(subsystem: "com.example.subsystem", category: "main")
    ///
    ///     let cancel = [1, 2, 3].publisher
    ///         .logEvents(to: logger, prefix: "Counter")
    ///         .sink { _ in } receiveValue: { _ in }
    ///
    ///      // sample output:
    ///      // 2021-03-16 20:41:10.980102+0200 subsystem[47288:1810254] [main] Counter started
    ///      // 2021-03-16 20:41:10.980252+0200 subsystem[47288:1810254] [main] Counter output
    ///      // 2021-03-16 20:41:10.980361+0200 subsystem[47288:1810254] [main] Counter output
    ///      // 2021-03-16 20:41:10.980451+0200 subsystem[47288:1810254] [main] Counter output
    ///      // 2021-03-16 20:41:10.980578+0200 subsystem[47288:1810254] [main] Counter finished
    /// - Parameters:
    ///   - logger: the optional logger to log to
    ///   - prefix: the prefixed text before any of the lifecycle methods
    /// - Returns: a publisher similar to the upstream publisher
    func logEvents(
        to logger: Logger?,
        prefix: String
    ) -> AnyPublisher<Output, Failure> {
        guard let logger = logger else { return self.handleEvents().eraseToAnyPublisher() }

        let receiveSubscription: ((Subscription) -> Void) = { _ in
            logger.log("\(prefix, privacy: .public) started")
        }

        let receiveCompletion: ((Subscribers.Completion<Self.Failure>) -> Void) = { completion in
            switch completion {
            case .failure(let error):
                logger.log(
                    level: .error,
                    "\(prefix, privacy: .public) error: \(error.localizedDescription, privacy: .public)"
                )
            case .finished:
                logger.log("\(prefix, privacy: .public) finished")
            }
        }

        let receiveOutput: ((Output) -> Void) = { _ in
            logger.log("\(prefix, privacy: .public) output")
        }

        let receiveCancel: (() -> Void) = {
            logger.log("\(prefix, privacy: .public) cancelled")
        }

        return self.handleEvents(
            receiveSubscription: receiveSubscription,
            receiveOutput: receiveOutput,
            receiveCompletion: receiveCompletion,
            receiveCancel: receiveCancel
        )
        .eraseToAnyPublisher()
    }
}
